### DISCUSSION TITLE
ERT Schlumberger synthetic model, in order to forward simulate and invert a landslide

### DISCUSSION THREAD

**Discussion by AY131313** [gimli-org/pyGIMLi#578] [General]
Hi, I'm trying to make a 3 layered subsurface landslide geometry. 
I provide the code I am running and would appreciate any insights and ideas to help me get through conducting a good inversion. 
The array I'm trying to use is Schlumberger. 

```
import matplotlib.pyplot as plt
import os
from pygimli.viewer.mpl import generateMatrix
from PIL import Image
from numpy import asarray
import numpy as np
from pygimli.meshtools import quality
import random 
import pygimli as pg
import pygimli.meshtools as mt
from pygimli.physics import ert

# Create geometry as provided
world = mt.createWorld(start=[0, 165], end=[117.5, 110])
layer1 = mt.createPolygon([[0.0, 137], [117.5, 164], [117.5, 162], [0.0, 135]],
                          isClosed=True, marker=1, area=1, boundaryMarker = 1)
layer2 = mt.createPolygon([[0.0, 126], [0.0, 135], [117.5, 162], [117.5, 153]],
                          isClosed=True, marker=2, boundaryMarker = 2)
layer3 = mt.createPolygon([[0.0, 110], [0.0, 126], [117.5, 153], [117.5, 110]],
                          isClosed=True, marker=3, boundaryMarker = 3)
freeAir = mt.createPolygon([[117.5,165] , [60,165] ,[0,165], [0,137]] , marker = 4, isClosed = True)
geom = layer1 + layer2 + layer3

# Define the number of electrodes and the desired positions
num_electrodes = 30
x_values = np.linspace(0.0, 117.5, num_electrodes)
y_values = np.linspace(137, 164, num_electrodes)
electrode_positions = np.column_stack((x_values, y_values))

# Create electrodes at the specified positions
for pos in electrode_positions:
    geom.createNode(pos[0], pos[1], 0.0)

ax, cb = pg.show(geom, showNodes=True)

#mesh creation
mesh = mt.createMesh(geom, quality = 32, area= 1)
pg.show(mesh);
plt.savefig("MODEL2.jpg",dpi=1200, bbox_inches="tight")


# Debugging: Check if any nodes have no cells
orphaned_nodes = []
for pos in electrode_positions:
    node = mesh.findNearestNode(pos)
    if node is None:
        orphaned_nodes.append(pos)

if orphaned_nodes:
    pg.warn("Orphaned nodes found:", len(orphaned_nodes))
    for i, pos in enumerate(orphaned_nodes):
        pg.warn(f"Orphaned Node {i+1}: Position {pos}")
        
        
# Create electrode data scheme
scheme = ert.createData(elecs=electrode_positions, schemeName="gr")
scheme['k'] = ert.geometricFactors(scheme) 
ert.show(scheme, 'k')
print(scheme)        


# Update electrode positions to match the mesh nodes
for i, pos in enumerate(scheme.sensorPositions()):
    nodeID = mesh.findNearestNode(pos)
    node = mesh.node(nodeID)
    electrode_positions[i] = [node.x(), node.y()]
pg.show(mesh)


# Debugging: Check if any nodes have no cells
orphaned_nodes = []
for pos in electrode_positions:
    node = mesh.findNearestNode(pos)
    if node is None:
        orphaned_nodes.append(pos)

if orphaned_nodes:
    pg.warn("Orphaned nodes found:", len(orphaned_nodes))
    for i, pos in enumerate(orphaned_nodes):
        pg.warn(f"Orphaned Node {i+1}: Position {pos}")
        
        
 #Define the resistivities for different regions using the region markers
resistivityMap = {
    1: 10.0,   # Secondary compacted regions with resistivity 10 ohm.m
    2: 75.0,   # Region with high moisture content with resistivity 75 ohm.m
    3: 150.0   # Hezar-Dareh alluvium with resistivity 150 ohm.m
}

# Assign resistivity values to the mesh using the region markers
resistivityValues = [resistivityMap[marker] for marker in mesh.cellMarkers()]

# Take a look at the mesh and the resistivity distribution
pg.show(mesh, data=resistivityValues, label=pg.unit('res'), showMesh=True)
plt.savefig("MODEL5.jpg",dpi=1200, bbox_inches="tight")


#forward simulation
data = ert.simulate(mesh, scheme=scheme, res=resistivityValues, noiseLevel = 0.03)
                    
pg.info(np.linalg.norm(data['err']), np.linalg.norm(data['rhoa']))
pg.info('Simulated data', data)
pg.info('The data contains:', data.dataMap().keys())

pg.info('Simulated rhoa (min/max)', min(data['rhoa']), max(data['rhoa']))
pg.info('Selected data noise %(min/max)', min(data['err'])*100, max(data['err'])*100)
```

after the last code snnipet related to forward simulation I always recieve an error like below:
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[25], line 1
----> 1 data = ert.simulate(mesh, scheme=scheme, res=resistivityValues)
      4 pg.info(np.linalg.norm(data['err']), np.linalg.norm(data['rhoa']))
      5 pg.info('Simulated data', data)

File E:\Anaconda3\envs\pg-transform2022\lib\site-packages\pygimli\physics\ert\ert.py:121, in simulate(mesh, scheme, res, **kwargs)
    118 # fop = self.createForwardOperator(useBert=True,  # self.useBert,
    119 #                                   sr=sr, verbose=verbose)
    120 fop.data = scheme
--> 121 fop.setMesh(mesh, ignoreRegionManager=True)
    123 rhoa = None
    124 phia = None

File E:\Anaconda3\envs\pg-transform2022\lib\site-packages\pygimli\frameworks\modelling.py:655, in MeshModelling.setMesh(self, mesh, ignoreRegionManager)
    652         super(Modelling, self).setMesh(mesh, ignoreRegionManager=True)
    653         pass
--> 655     self.setMeshPost(mesh)
    656     return
    657 self.clearRegionProperties()

File E:\Anaconda3\envs\pg-transform2022\lib\site-packages\pygimli\physics\ert\ertModelling.py:185, in ERTModelling.setMeshPost(self, mesh)
    183 def setMeshPost(self, mesh):
    184     """"""
--> 185     self._core.setMesh(mesh, ignoreRegionManager=True)

RuntimeError: C:/msys64/home/halbm/gimli/gimli/core/src/mesh.cpp:785		GIMLI::Cell* GIMLI::Mesh::findCell(const RVector3&, size_t&, bool) const  no cells for this node. This is a corrupt mesh

P.S. I provide the pictures of the initial geometry, the Mesh, the data scheme, and the geometry merged with the mesh when injected resistivity values using the resistivity Map. 
![mesh+res values](https://github.com/gimli-org/gimli/assets/132713408/743b87fe-9c32-4bd6-9755-1fd8c603c426)
![data scheme](https://github.com/gimli-org/gimli/assets/132713408/a7df878e-e925-494c-bff7-e5b2e91ad208)
![Mesh](https://github.com/gimli-org/gimli/assets/132713408/4a2bc3b6-eb29-48d2-95c6-e98d2de67bb0)
![geometry with electrodes](https://github.com/gimli-org/gimli/assets/132713408/b63cf430-72fc-4e82-8956-6427c6f46952)

**Comment by halbmy** [comment]
I am getting an error already when I add the electrodes and want to show (i.e. in `pg.show`).

Plotting the mesh and the electrodes before (`pg.show(geo)[0].plot(x_values, y_values)`)

![image](https://github.com/gimli-org/gimli/assets/7136401/e040320e-de44-4a2b-992a-15d87b3c04ec)

it becomes clear that there are several issues
* the electrodes should not be on the side boundary
* some boundary around the electrodes is needed for boundary conditions of the forward problem
* creating a node where there is already one should be avoided by `createNodeWithCheck`
* there can be tolerance issues, i.e. the nodes are very close but below or above the topography line

I therefore suggest creating the mesh from the electrode positions so that they are included, using meshtools functions as done in several examples.

**Comment by halbmy** [comment]
Any progress?

**Comment by AY131313** [comment]
Hello Dr. Gunther.
Actually yes. I get an answer. I will mail it .I really appreciate your
help and support.
An honor to be in touch with you.
Regards.
Amir Yazdanpanah

**Comment by AY131313** [comment]
Now Im trying to add a topography to my final inversion results. Any ideas
I can start with??

**Comment by halbmy** [comment]
You should start right with the correct geometry, including the topography.

**Comment by AY131313** [comment]
You mean by creating a polygon?

**Comment by halbmy** [comment]
By creating a mesh that already contains the correct topography, and also by using it in the data (from which the mesh can be created).

**Comment by AY131313** [comment]
can you please elaborate a little more? I mean how can I get provided with
a mesh that already contains topography?

**Comment by AY131313** [comment]
Hello Dr.Gunther, 
I have made my ERT and SRT models. 
for inverting the ERT field data, I use pybert (pb.importData()).
in the inversion step my model is weird and not representable of my data (That belongs to a landslide and contains topography).
![Screenshot 2023-11-09 185620](https://github.com/gimli-org/gimli/assets/132713408/dfdbe705-6e19-4bae-bedc-a220740a8b79)
![Screenshot 2023-11-09 185918](https://github.com/gimli-org/gimli/assets/132713408/e12a4201-7de6-496b-817d-22d1c7b637eb)
![Screenshot 2023-11-09 190005](https://github.com/gimli-org/gimli/assets/132713408/8a994de4-37e1-4cd7-b4de-208fe0df3108)
![Screenshot 2023-11-09 190043](https://github.com/gimli-org/gimli/assets/132713408/b462f65b-1c07-47ec-892f-fd37209d2172)
![Screenshot 2023-11-09 190110](https://github.com/gimli-org/gimli/assets/132713408/b585507a-3248-453b-b6c0-bfde432ff566)

**Comment by halbmy** [comment]
1. Which format does your data have? Have you tried `ert.load()`?
2. It seems like the topography is not read properly (numerical are identical with analytical geometric factors, not topo in results).
3. Your pseudosection looks really weird? Why?

**Comment by AY131313** [comment]
hello. i used a .dat file in the format of res2dinv.
yes I did but ert.load() didn't read my data. 
because it belongs to a landslide and doesn't represent it properly.

**Comment by halbmy** [comment]
`ert.load()` should be able to read res2dinv format and uses exactly the same function as pybert importData. Maybe we made an error, would be good to check if you can share the data.

**Comment by AY131313** [comment]
Hello Dr,
I tried ert.load() and I got an inversion result without any topography (the case is related to detecting the sliding surface of a landslide).
I have sent the field data of a profile in res2dinv format to your academic email. 
![ERTfield](https://github.com/gimli-org/gimli/assets/132713408/fffa622a-7d06-42a4-a383-7a10aa854108)
![firstERTrealdata](https://github.com/gimli-org/gimli/assets/132713408/ad941d4f-b2c8-4890-9540-cbf75a55a67d)

**Comment by halbmy** [comment]
Something's wrong with the data import. I can't tell unless seing the data.

**Reply by AY131313** [reply]
the data has been sent to your email address Dr.

**Comment by halbmy** [comment]
I haven’t got anything. I haven’t even an idea who you are.

Von: AY13 ***@***.***>
Gesendet: Donnerstag, 16. November 2023 21:49
An: gimli-org/gimli ***@***.***>
Cc: Guenther, Thomas ***@***.***>; Comment ***@***.***>
Betreff: Re: [gimli-org/gimli] ERT shclumberger synthetic model, in order to forward simulate and invert a landslide (Discussion #578)


the data has been sent to your email address Dr.

**Reply by AY131313** [reply]
![Screenshot_20231117-152529_Gallery](https://github.com/gimli-org/gimli/assets/132713408/fce76ef4-1ac0-47a0-ba31-c6c954ba5308)

**Comment by AY131313** [comment]
Hello dr, 
any progress here? I mean, have u found the data I emailed to you?
If not, please send me an email address.
regards.

**Comment by halbmy** [comment]
As this discussion moved out of the public, I am closing it.
